File: android/app/src/main/java/com/dailysns/viewmodels/FeedViewModel.kt
package com.dailysns.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dailysns.models.Post
import com.dailysns.services.DataService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * FeedViewModel.kt
 * DailySNSWithYourSS Android版
 * 
 * フィード表示のビジネスロジック
 */

enum class FeedType {
    FOLLOWING,
    GLOBAL
}

class FeedViewModel(
    private val currentUserId: String,
    private val dataService: DataService = DataService()
) : ViewModel() {
    
    private val _posts = MutableStateFlow<List<Post>>(emptyList())
    val posts: StateFlow<List<Post>> = _posts.asStateFlow()
    
    private val _feedType = MutableStateFlow(FeedType.FOLLOWING)
    val feedType: StateFlow<FeedType> = _feedType.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    /**
     * 投稿を取得
     */
    fun fetchPosts() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            
            try {
                _posts.value = when (_feedType.value) {
                    FeedType.FOLLOWING -> dataService.fetchRandomFollowingPosts(currentUserId)
                    FeedType.GLOBAL -> dataService.fetchRandomGlobalPosts()
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    /**
     * フィードタイプを切り替え
     */
    fun switchFeedType(type: FeedType) {
        if (_feedType.value == type) return
        _feedType.value = type
        fetchPosts()
    }
    
    /**
     * 更新
     */
    fun refresh() {
        fetchPosts()
    }
}

File: web/src/viewmodels/FeedViewModel.ts
/**
 * FeedViewModel.ts
 * DailySNSWithYourSS Web版
 * 
 * フィード表示のビジネスロジック
 */

import { Post } from '../models/Post';
import { DataService } from '../services/DataService';

export enum FeedType {
  Following = 'following',
  Global = 'global'
}

export class FeedViewModel {
  private _posts: Post[] = [];
  private _feedType: FeedType = FeedType.Following;
  private _isLoading: boolean = false;
  private _errorMessage: string | null = null;
  private listeners: Set<() => void> = new Set();
  private dataService: DataService;
  private currentUserId: string;

  constructor(currentUserId: string, dataService: DataService = new DataService()) {
    this.currentUserId = currentUserId;
    this.dataService = dataService;
  }

  get posts(): Post[] {
    return this._posts;
  }

  get feedType(): FeedType {
    return this._feedType;
  }

  get isLoading(): boolean {
    return this._isLoading;
  }

  get errorMessage(): string | null {
    return this._errorMessage;
  }

  /**
   * 変更を監視
   */
  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notify(): void {
    this.listeners.forEach(listener => listener());
  }

  /**
   * 投稿を取得
   */
  async fetchPosts(): Promise<void> {
    this._isLoading = true;
    this._errorMessage = null;
    this.notify();

    try {
      if (this._feedType === FeedType.Following) {
        this._posts = await this.dataService.fetchRandomFollowingPosts(this.currentUserId);
      } else {
        this._posts = await this.dataService.fetchRandomGlobalPosts();
      }
    } catch (error) {
      this._errorMessage = error instanceof Error ? error.message : 'Unknown error';
    } finally {
      this._isLoading = false;
      this.notify();
    }
  }

  /**
   * フィードタイプを切り替え
   */
  async switchFeedType(type: FeedType): Promise<void> {
    if (this._feedType === type) return;
    this._feedType = type;
    await this.fetchPosts();
  }

  /**
   * 更新
   */
  async refresh(): Promise<void> {
    await this.fetchPosts();
  }
}
